### DISCUSSION TITLE
github.event.pull_request.base.sha should be updated each time the pull request is updated

### DISCUSSION THREAD

**Discussion by JavaScriptBach** [community/community#59677] [Actions]
### Select Topic Area

Product Feedback

### Body

This has been discussed a bit in https://github.com/orgs/community/discussions/39880 and https://github.com/actions/runner/issues/1689, but the discussions there have gone a bit on a tangent. I wanted to make a separate thread because I recently spent a lot of time debugging this behavior in my CI pipeline and I strongly believe the current behavior, while [technically consistent with documentation](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request), results in an subpar developer experience and deserves to be changed. Let me make the case:

# Current Behavior
The value of `${{ github.event.pull_request.base.sha }}` is not updated on every commit to the PR. It is the SHA of the base branch at the time the PR was created, which becomes out of date as the PR gets updated.

# Desired Behavior
The value of `${{ github.event.pull_request.base.sha }}` should be updated on every commit to the value of the base branch SHA used to construct the PR's merge branch.

# Why
A core use case when writing GitHub Actions workflows on pull requests is to find the changes introduced in the PR. This requires finding 2 commits that will produce a correct diff output, which can then be passed to `git diff` or various other third-party apps and integrations.

If we choose the PR branch as the head, (i.e. `${{ github.event.pull_request.head.sha }}`), then the correct base SHA is `$(git merge-base ${{ github.event.pull_request.base.sha }} ${{ github.event.pull_request.head.sha }})`. This produces the equivalent 3-dot-diff behavior that the GitHub UI uses.

If we choose the PR's merge branch as the head (i.e. `${{ github.sha }}`), then the correct base SHA is the value of the base branch used to construct the merge branch. This is currently NOT `${{ github.event.pull_request.base.sha }}`, because that value isn't updated every time the PR is updated.  The only way I can currently get this base SHA is to parse the auto-generated commit message in the merge branch. (They currently have a message of "Merge \<head sha\> into \<base sha\>"). The ask is to put \<base sha\> as the value of `${{ github.event.pull_request.base.sha }}` so I can use the latter instead.

In general, I think **there should be a first-class way of getting the base SHA where the head is the PR's _merge branch_**. This is because actions on PRs [run on the PR's merge branch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request), not the PR branch. Because PR actions run on the merge branch, it's easiest for the workflow to also check out the merge branch and run checks on it. Checking out any other branch results in version skew between the workflow version and the repo, leading to a subpar developer experience.

**Comment by afinetooth** [comment]
### Upvote
I want to upvote this. 

I see the problem from a different angle, but the issue remains the inconsistency between the [documented principle](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#git-diff-comparisons:~:text=GitHub%20generates%20the%20list%20of%20changed%20files%20using%20two%2Ddot%20diffs%20for%20pushes%20and%20three%2Ddot%20diffs%20for%20pull%20requests%3A) that [GitHub uses three-dot diffs for pull requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github), and the fact that Github Actions' [`pull_request` builds  reflect a two-dot diff comparison](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#two-dot-git-diff-comparison).

I'll pose it a different way if that's not clear:

- IF 
  - [Github generates **three-dot diffs** for pull requests](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#git-diff-comparisons:~:text=GitHub%20generates%20the,the%20base%20branch.) ([focused on "_what a pull request introduces_"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github:~:text=Since%20the%20three%2Ddot%20comparison%20compares%20with%20the%20merge%20base%2C%20it%20is%20focusing%20on%20%22what%20a%20pull%20request%20introduces%22.))
- AND
- IF 
  - [The **three-dot diff** comparison shows the difference between the **latest common commit** of both branches (merge base) and the most recent version of the topic branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#three-dot-git-diff-comparison).
- THEN
  - Why are **`pull_request` _builds_** in Github Actions predicated on [**two-dot diff comparisons**](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#two-dot-git-diff-comparison) (which show all changes on both base and topic branches)?

I understand that, as a CI service, Github Actions is _primarily_ concerned with "Continuous Integration," and that, as [this user](https://github.com/ethomson) says [here](https://github.com/actions/checkout/issues/27#issuecomment-535897113):

> Continuous integration builds need to take into account what they'll be merge into, not the state of the repository when they were created. This prevents you from merging a commit that breaks master but "worked on my machine".

However, it just leads to so much confusion ([ex1](https://github.com/actions/checkout/issues/504), [ex2](https://github.com/actions/checkout/issues/27) ...) that this is the _default_ behavior.

It feels like there should, at least, be a clearer way to specify what you want from a `pull_request` build: a _snapshot build_ for the pull request, or a _merge-ability report_ for it.

### Our use case:

I work for a service in the CI toolchain that tracks test coverage for a repo's successive commits, potentially for its full history. 

Where this test coverage is _most_ useful is for `pull_request` builds, since users can choose to do things like block PRs from being merged when coverage falls below a certain %, etc.

The problem, for us, is that to determine changes in coverage between two commits, we need to diff their _coverage reports_, which are predicated on a _build_ for each commit. 

For a given pull request, we've typically already received a coverage report for the `push` build on the target branch that represents its `base` commit.

However, since the `pull_request` build from Github Actions (by default, based on the [merge commit](https://blog.mergify.com/what-is-the-difference-between-a-merge-commit-a-squash/#:~:text=Git%20tracks%20commits,have%20checked%20out.)) includes any recent changes on the base branch, we get a coverage report that _includes_ those changes, which can't be diff'd against the report for the `base` commit.

But that's the coverage change the user expects to see. They want to see how coverage has changed for _their_ pull request&mdash;to wit, for "[_what [their] pull request introduces_](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github:~:text=Since%20the%20three%2Ddot%20comparison%20compares%20with%20the%20merge%20base%2C%20it%20is%20focusing%20on%20%22what%20a%20pull%20request%20introduces%22.)"&mdash;so they understand what changes they can make to _their_ pull request, to _improve_ coverage.

**Comment by carlescufi** [comment]
As of September 2024,  `pr.base.sha` always corresponds to the SHA of the head/tip of the base branch at the moment of either creation of the PR **or** the last **force-push** (regardless of whether that force-push actually rebased the PR branch against base or was just modifying an existing commit in it) to the PR branch. This makes sense since the only way to rebase against the latest base branch requires force-pushing to the PR branch.

Hence, this means that:

`git merge-base <pr.base.sha> <pr.head.sha>`

will always yield the correct fork point of the PR branch off the base branch.

This in turn means that one can list all commits in a PR with:

`git rev-list <pr.base.sha>..<pr.head.sha>`

or its equivalent:

`git rev-list ^<pr.base.sha> <pr.head.sha>`

**Comment by JavaScriptBach** [comment]
@carlescufi this doesn't fix the issue. The ask is that the value of `github.event.pull_request.base.sha` be the value of the base branch used to construct the merge branch. I just tested this and it is not the case in my Actions workflow.

Merge branch:
```
HEAD is now at 9a02cc48 Merge 3a393bb1daf445ffac4cc892ba36317f4424a092 into bdd12cacd1acacce2b37138b01d8167da53bb0fb
```

`github.event.pull_request.base.sha` and `github.event.pull_request.head.sha`:
```
Pull request head: 3a393bb1daf445ffac4cc892ba36317f4424a092
Pull request base: f7179ac29fd5d0f756eaa3dc3fb129428aab2b8e
```

**Reply by carlescufi** [reply]
I see @JavaScriptBach, sorry for the confusion. Perhaps I did not understand the issue you are facing correctly then. I just reread the GitHub Doc on the PR "merge branch", I honestly wasn't aware that there was a merge commit involved at all. In my case when I run a GitHub action it runs on a tree with no merge commits at all, that is, it runs on whatever tree is present in the branch used to open the Pull Request. I am unsure as to why you see a merge commit, I guess you are not using a tree checkout of the pull request's at `pr.head.sha`, which is what I use.

I have now looked a bit more into this. The reason I never see the merge commit you refer to is that I always use the checkout action like so:

```
    - name: checkout
      uses: actions/checkout@v4
      with:
        ref: ${{ github.event.pull_request.head.sha }}
```

The use of `github.event.pull_request.head.sha` means that only the commits actually present in the PR branch (i.e. the actual branch the PR author created) are visible. If I need to update the tip of the PR branch with the latest base branch then I do:

`git rebase origin/${{ github.base_ref }}`

When used in this way, then you always have a clean branch which is equivalent to the base branch with that PR's commits applied on top.

**Reply by carlescufi** [reply]
I have now tried locally as well, to understand this better.

If I push additional commits to both the base and PR branches (after creating the PR),  I get the difference:

```
base.sha: 2002444ece71971f03bc9068978d15386aafa6f0
head.sha: 58e0feff9e60356[21](https://github.com/nrfconnect/sdk-nrf-testing/actions/runs/11127340380/job/30919418678?pr=262#step:4:22)13164941911f2be2c600093
74539365c1ba007fefb0a82fd4a89184741f85c1 Merge 58e0feff9e603562113164941911f2be2c600093 into 8c3d88a341794b0b5dbd6aa1789ef08a54d3a5ce
```

But in my opinion this is intentionate. The `base.sha` still points correctly to the tip of the base branch, and will only be updated if you force-push to the PR branch, as described in my comment [here](https://github.com/orgs/community/discussions/59677#discussioncomment-10782359). However, the merge commit is trying to merge the PR branch against the latest base branch, to ensure that there are no merge conflicts.

If you need to recreate the pure, "double dot" list of commits (git log)  you can simply use:

`git rev-list <pr.base.sha>..<pr.head.sha>`

or its equivalent:

`git rev-list ^<pr.base.sha> <pr.head.sha>`

To see the pure, "triple dot" diff, this will always do the right thing:
`git diff <pr.base.sha>...<pr.head.sha>`

regardless of what you have checked out, this should always give you what you need, if I understand this correctly.

**Reply by JavaScriptBach** [reply]
I use ${{ github.sha }}.

**Reply by carlescufi** [reply]
Right, so if you use `${{ github.sha }}` then I don't think that GitHub will give you what you want. `base.sha`, from what I can tell, rightly points to "the fork point of the PR branch off of the base branch (aka merge base)" and will only ever be updated in a PR if you **force-push** to the PR branch.  This makes sense, because unless you force-push you cannot change the merge base at all. But since GitHub needs to check if a branch is mergeable against whatever the latest tip of the base branch is, it will use that tip instead of `base.sha` in the artificial merge commit it introduces. 

You could therefore switch to using `head.sha` instead (and optionally rebasing against the latest base branch) or you could, instead of parsing the merge commit message, use Git to find SHA you are after from the merge commit: 

`git rev-parse ${{github.sha}}^1`, which points to the first parent of the merge commit, which is always the tip of the base branch.

**Reply by JavaScriptBach** [reply]
I can't use head.sha because of the reasons I mentioned in the last paragraph of my post. It sounds like GitHub has not changed any behavior here so the issue is still open.

**Reply by carlescufi** [reply]
Right, but when you write:

>This is because actions on PRs [run on the PR's merge branch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request), not the PR branch.

what do you mean? you choose what ref to check out when you actually check out the code  in the workflow.

**Reply by afinetooth** [reply]
@carlescufi apologies if you know this already. I find myself explaining it a lot to new users of our service, so for anyone who may be reading this who's not super familiar with git and how `pull_request` commits work...

`pull_request` commits are built on what's known as the "merge branch," which isn't really a branch but can be thought of as a _temporary_ branch that's separate from the rest of the branches on your git tree. 

That's because `pull_request` commits are _prospective_ commits. They're not "historical" commits, like `push` commits, in the sense that they won't live forever in your git tree / history. They are created to show "what would happen if you merged the HEAD of your PR branch into the HEAD of your target branch, right now," and provide a basis for deciding whether to _actually_ merge the PR branch into the target branch. 

And then, once the PR branch _is_ merged into the target branch, git creates a new, _historical_ `push` commit on your target branch.

That's how I think of it, anyway. I'm sure there's a more technically accurate definition.

> you choose what ref to check out when you actually check out the code in the workflow.

Not really. One of the HEAD commits comes from your PR branch, which is why you may see the PR branch mentioned. It's the _context_ of the `pull_request` commit. But the commit itself is built on the merge branch, and you can tell that's happening when you see an (automated) commit message like "Merge 2c8d25231 into 42ed6119a."

**Reply by carlescufi** [reply]
@afinetooth thanks for the additional context

>This is because pull_request commits are prospective commits. They're not really historical commits, like push commits, in the sense that they don't live forever in your git tree / history. They are created to show "what would happen if you merged the HEAD of your PR branch into the HEAD of your target branch, right now," and provide the basis for deciding whether to actually merge the PR branch into the target branch, which, when done, creates a new, historical push commit on your target branch.

Yep, I do understand that, though please note that you can also "merge" the PR by rebasing the commits in the PR branch, which then creates no merge commits at all, and allows you to have a pure linear history.

>Not really. One of the HEAD commits comes from your PR branch, which is why you may see the PR branch mentioned. It's the context of the pull_request commit. But the commit itself is built on the merge branch, and you can tell that's happening when you see an (automated) commit message like "Merge 2c8d25231 into 42ed6119a."

But you **choose** what code to check out in order to run tests on your Pull Request. As I wrote [here](https://github.com/orgs/community/discussions/59677#discussioncomment-10808053), I always check out `head.sha` before doing anything else (such as running tests, checking diffs against the base branch, etc). The fact that GitHub provides this merge commit does not mean you need to use it at all or check it out ever. In some of the bigger projects where I contribute this merge commit, and in fact the "virtual" merge branch are basically completely ignored and never actuallly used in the actions/workflows. We always check the tip of the PR branch (`head.sha`) and then, if needed, rebase it against the tip of the base branch.

**Reply by JavaScriptBach** [reply]
> Right, but when you write:
> 
> > This is because actions on PRs [run on the PR's merge branch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request), not the PR branch.
> 
> what do you mean? you choose what ref to check out when you actually check out the code in the workflow.

I choose what ref to check out, but I don't choose what version of the workflow that runs. That version is always github.sha. Choosing any ref other than that leads to version skew.

**Reply by carlescufi** [reply]
> > Right, but when you write:
> > > This is because actions on PRs [run on the PR's merge branch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request), not the PR branch.
> > 
> > 
> > what do you mean? you choose what ref to check out when you actually check out the code in the workflow.
> 
> I choose what ref to check out, but I don't choose what version of the workflow that runs. That version is always github.sha. Choosing any ref other than that leads to version skew.

If you refer to the revision of the workflow file itself that is being used to execute the action, that depends on whether you use `pull_request` or `pull_request_target`. That said, I am not quite sure how that is important other than for PRs that actually modify the workflow file itself, but I'm sure there are good reasons.

**Comment by antspy** [comment]
Hi, 

Any update on this? Is there a better way to understand the behavior of merges into the main branch at ci time ?
